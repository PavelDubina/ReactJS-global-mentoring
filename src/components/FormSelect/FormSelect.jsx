import React, { useState, useRef } from 'react'
import PropTypes from 'prop-types'
import className from 'classnames'
import { useOutsideClick } from '../../Hooks/useClickOutside'
import { genres } from '../../utils/constants'
import styles from './FormSelect.scss'

export const FormSelect = ({ name, value, onChange }) => {
  const ref = useRef()
  const [isOpen, setIsOpen] = useState(false)
  const dropStyle = className(styles.dropdown, { [styles.visible]: isOpen })
  const arrowStyle = className(styles.arrow, { [styles.open]: isOpen })
  const valueStyle = className(styles.value, { [styles.default_value]: !value })
  useOutsideClick(ref, () => setIsOpen(false))
  return (
    <>
      <span>GENRE</span>
      <div ref={ref} className={styles.container}>
        <div onClick={() => setIsOpen((state) => !state)} className={styles.select}>
          <p className={valueStyle}>{value || 'Select genre'}</p>
          <div className={arrowStyle}>&#9660;</div>
        </div>
        <div data-testid="dropdown" className={dropStyle}>
          {genres.map((genre) => (
            <div key={genre} className={styles.input}>
              <input
                className={styles.checkbox}
                name={name}
                id={genre}
                type="checkbox"
                checked={value.split(', ').includes(genre)}
                value={genre}
                onChange={onChange}
              />
              <label className={styles.lable} htmlFor={genre}>
                <span>{genre}</span>
              </label>
            </div>
          ))}
        </div>
      </div>
    </>
  )
}

FormSelect.propTypes = {
  value: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
}
