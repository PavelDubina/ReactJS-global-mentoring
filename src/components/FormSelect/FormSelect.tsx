import React, { useState, useRef } from 'react'
import className from 'classnames'
import { useTranslation } from 'next-i18next'
import { useOutsideClick } from '../../Hooks/useClickOutside'
import { genres } from '../../utils/constants'
import styles from '../../../styles/FormSelect.module.scss'

type FormSelectProps = {
  name: string
  value: string
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void
}

export const FormSelect: React.FC<FormSelectProps> = ({ name, value, onChange }) => {
  const { t } = useTranslation()
  console.log(t('addModal.genres.comedy'))
  const ref = useRef(null)
  const [isOpen, setIsOpen] = useState<boolean>(false)
  const dropStyle = className(styles.dropdown, { [styles.visible]: isOpen })
  const arrowStyle = className(styles.arrow, { [styles.open]: isOpen })
  const valueStyle = className(styles.value, { [styles.default_value]: !value })
  useOutsideClick(ref, () => setIsOpen(false))
  return (
    <>
      <span>{t('addModal.input.genre')}</span>
      <div ref={ref} className={styles.container}>
        <div onClick={() => setIsOpen((state) => !state)} className={styles.select}>
          <p className={valueStyle}>
            {(value.length > 0 &&
              value
                .split(', ')
                .map((genre) => t(`addModal.genres.${genre}`))
                .join(', ')) ||
              t('addModal.placeholders.genre')}
          </p>
          <div className={arrowStyle}>&#9660;</div>
        </div>
        <div data-testid="dropdown" className={dropStyle}>
          {genres.map((genre) => (
            <div key={genre} className={styles.input}>
              <input
                className={styles.checkbox}
                name={name}
                id={genre}
                type="checkbox"
                checked={value.split(', ').includes(genre)}
                value={genre}
                onChange={onChange}
              />
              <label className={styles.lable} htmlFor={genre}>
                <span>{t(`addModal.genres.${genre}`)}</span>
              </label>
            </div>
          ))}
        </div>
      </div>
    </>
  )
}
