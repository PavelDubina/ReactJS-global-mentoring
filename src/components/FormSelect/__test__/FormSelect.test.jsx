import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { FormSelect } from '../FormSelect'

describe('FormSelect', () => {
  const getComponent = ({ value = '', name = 'genres', onChange = jest.fn() }) =>
    render(<FormSelect value={value} name={name} onChange={onChange} />)
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Should open dropdown', () => {
    const { getByText, getByTestId } = getComponent({})
    const select = getByText('Select genre')
    const dropdown = getByTestId('dropdown')
    fireEvent.click(select)
    expect(dropdown.classList.contains('visible')).toBeTruthy()
  })
  it('Should close dropdown', () => {
    const { getByText, getByTestId } = getComponent({})
    const select = getByText('Select genre')
    const dropdown = getByTestId('dropdown')
    fireEvent.click(select)
    fireEvent.click(document)
    expect(dropdown.classList.contains('visible')).toBeFalsy()
  })
  it('Should call onChange', () => {
    const onChange = jest.fn()
    const { getByText, getByLabelText } = getComponent({ onChange })
    const select = getByText('Select genre')
    const checkbox = getByLabelText('comedy')
    fireEvent.click(select)
    fireEvent.click(checkbox)
    expect(onChange).toBeCalled()
  })
})
