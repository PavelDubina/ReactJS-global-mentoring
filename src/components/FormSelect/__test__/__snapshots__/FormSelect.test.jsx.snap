// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect Should renders correctly 1`] = `
<DocumentFragment>
  <span>
    GENRE
  </span>
  <div
    class="container"
  >
    <div
      class="select"
    >
      <p
        class="value default_value"
      >
        Select genre
      </p>
      <div
        class="arrow"
      >
        â–¼
      </div>
    </div>
    <div
      class="dropdown"
      data-testid="dropdown"
    >
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="action"
          name="genres"
          type="checkbox"
          value="action"
        />
        <label
          class="lable"
          for="action"
        >
          <span>
            action
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="adventure"
          name="genres"
          type="checkbox"
          value="adventure"
        />
        <label
          class="lable"
          for="adventure"
        >
          <span>
            adventure
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="animation"
          name="genres"
          type="checkbox"
          value="animation"
        />
        <label
          class="lable"
          for="animation"
        >
          <span>
            animation
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="comedy"
          name="genres"
          type="checkbox"
          value="comedy"
        />
        <label
          class="lable"
          for="comedy"
        >
          <span>
            comedy
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="crime"
          name="genres"
          type="checkbox"
          value="crime"
        />
        <label
          class="lable"
          for="crime"
        >
          <span>
            crime
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="documentary"
          name="genres"
          type="checkbox"
          value="documentary"
        />
        <label
          class="lable"
          for="documentary"
        >
          <span>
            documentary
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="drama"
          name="genres"
          type="checkbox"
          value="drama"
        />
        <label
          class="lable"
          for="drama"
        >
          <span>
            drama
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="family"
          name="genres"
          type="checkbox"
          value="family"
        />
        <label
          class="lable"
          for="family"
        >
          <span>
            family
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="fantasy"
          name="genres"
          type="checkbox"
          value="fantasy"
        />
        <label
          class="lable"
          for="fantasy"
        >
          <span>
            fantasy
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="history"
          name="genres"
          type="checkbox"
          value="history"
        />
        <label
          class="lable"
          for="history"
        >
          <span>
            history
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="horror"
          name="genres"
          type="checkbox"
          value="horror"
        />
        <label
          class="lable"
          for="horror"
        >
          <span>
            horror
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="music"
          name="genres"
          type="checkbox"
          value="music"
        />
        <label
          class="lable"
          for="music"
        >
          <span>
            music
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="mystery"
          name="genres"
          type="checkbox"
          value="mystery"
        />
        <label
          class="lable"
          for="mystery"
        >
          <span>
            mystery
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="romance"
          name="genres"
          type="checkbox"
          value="romance"
        />
        <label
          class="lable"
          for="romance"
        >
          <span>
            romance
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="science fiction"
          name="genres"
          type="checkbox"
          value="science fiction"
        />
        <label
          class="lable"
          for="science fiction"
        >
          <span>
            science fiction
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="thriller"
          name="genres"
          type="checkbox"
          value="thriller"
        />
        <label
          class="lable"
          for="thriller"
        >
          <span>
            thriller
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="war"
          name="genres"
          type="checkbox"
          value="war"
        />
        <label
          class="lable"
          for="war"
        >
          <span>
            war
          </span>
        </label>
      </div>
      <div
        class="input"
      >
        <input
          class="checkbox"
          id="western"
          name="genres"
          type="checkbox"
          value="western"
        />
        <label
          class="lable"
          for="western"
        >
          <span>
            western
          </span>
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
